### PR TITLE
Foolproof hotend offset editing

### PR DESCRIPTION
**Pull request by InsanityAutomation**
Opening a little early as it has an overlapping fix with @Ludy87

Hasnt been tested on hardware yet, thats tomorrows plan! I also plan to add some logic around the z softendstop and hotend offset clamp similar to the probe minimum allowance to go below.

@thinkyhead This contains the flip in the section of logic that it looked to me like the condition was inverted in motion.cpp for you to review.

### CODE REVIEW

**Comment by Ludy87** [comment]
@InsanityAutomation

then ignore my PR and use your branch, ok?

**Comment by InsanityAutomation** [comment]
@Ludy87 Not really a big deal either way to me, its up to Scott but figured id at least show what I meant since I had already made the same change offline awhile ago. All this really does differently from yours is add the lines in after config_adv is processed instead of before.

**Comment by Ludy87** [comment]
Just wanted to agree with your change, Scott knows my Rush PRs 🤣

**Comment by InsanityAutomation** [comment]
@Ludy87 LOL sounds good. I usually try to test everything on my end so sometimes things sit for a few days before I open a PR with em.... Unless im in a hurry on something ;-P

File: Marlin/src/module/motion.cpp [hunk 1]
```diff
@@ -573,26 +573,35 @@ void clean_up_after_endstop_or_probe_move() {
 
     #else
 
-      #if !HAS_SOFTWARE_ENDSTOPS || ENABLED(MIN_SOFTWARE_ENDSTOP_X)
+      #if HAS_SOFTWARE_ENDSTOPS || ENABLED(MIN_SOFTWARE_ENDSTOP_X)
```
Comment (thinkyhead) [inline Marlin/src/module/motion.cpp]: This is breaking the intended logic!

**Comment by thinkyhead** [comment]
Rebased.
```
git fetch origin
git checkout OffsetClamping
git reset --hard origin/OffsetClamping
```

File: Marlin/src/module/motion.cpp (hunk 1 above)
Comment (InsanityAutomation) [reply Marlin/src/module/motion.cpp]: So its intended to clamp to a soft limit even when soft limits are disabled?

Comment (thinkyhead) [reply Marlin/src/module/motion.cpp]: When they don't exist, the array is filled in with defaults.

Comment (InsanityAutomation) [reply Marlin/src/module/motion.cpp]: Fair enough, like I said when I passed through this briefly last night it just stuck out and I hadnt looked at it much yet. I would have expected no restriction if disabled. You see my reasoning for posting this before I had a chance to dig more above.

Comment (thinkyhead) [reply Marlin/src/module/motion.cpp]: A hint that it was intended: Any of the `ENABLED(Mxx_SOFTWARE_ENDSTOP_x)` will result in `HAS_SOFTWARE_ENDSTOPS` being true.

**Comment by InsanityAutomation** [comment]
Ill add clamping for babystepping tomorrow...

**Comment by InsanityAutomation** [comment]
Last comit there is just an idea of where im going. Havnt even made sure x/y hotend offset application is in the right direction.... Out of time right now but figured id let you see where im going.

**Comment by boelle** [comment]
@thinkyhead this one is failing too

**Comment by InsanityAutomation** [comment]
@boelle this one was mainly a concept to discuss not a completed code change. I'll circle back to it in awhile

**Comment by boelle** [comment]
hehe, i just saw the fail and wanted to raise some attention keep the hard working people without enough time on their toes :-D

**Comment by thinkyhead** [comment]
I like the concept of doing a simple interface on the LCD where you align each tool to a spot on the bed, and then set it that way. In the same process you could also tune the Z probe offset.

**Comment by InsanityAutomation** [comment]
> 
> 
> I like the concept of doing a simple interface on the LCD where you align each tool to a spot on the bed, and then set it that way. In the same process you could also tune the Z probe offset.

Originally this was looked at as the first step in helpers to bring M425 into function for IDEX which would do exactly that. Alas priority shifted for interested parties so my efforts have been sidetracked for the time being.

**Comment by thinkyhead** [comment]
Did rebase, squash, clean up, and modernize…
```
git fetch origin
git checkout OffsetClamping
git reset --hard origin/OffsetClamping
```

File: Marlin/src/gcode/motion/M290.cpp [hunk 2]
```diff
@@ -40,7 +40,44 @@
...
+        if (offs > hotend_offset[active_extruder].x + hotend_offset_limit.x)
+          offs = -hotend_offset_limit.x - zprobe_zoffset;
+        else if (offs < hotend_offset[active_extruder].x - hotend_offset_limit.x)
+          offs = hotend_offset_limit.x - hotend_offset[active_extruder].x;
+      }
+      else if (axis == Y_AXIS) {
+        if (offs > hotend_offset[active_extruder].y + hotend_offset_limit.y)
+          offs = -hotend_offset_limit.y - zprobe_zoffset;
+        else if (offs < hotend_offset[active_extruder].y - hotend_offset_limit.y)
+          offs = hotend_offset_limit.y - hotend_offset[active_extruder].y;
```
Comment (thinkyhead) [inline Marlin/src/gcode/motion/M290.cpp:74]: I rearranged the comparisons to put `offs` in front of the compare operator. Please validate these changes.

**Comment by boelle** [comment]
the idea of doing offsets to a spot on the bed sounds nice :-D

**Comment by boelle** [comment]
should this not have been merged or does it still need some work?

**Pull request closed without merging**